man/bulkrec.3
.TH BULKREC 3 "12 September 1995"
.FB
.SH NAME
bulkrec_begin, bulkrec_end \- bulk record process control
.SH SYNOPSIS
#include <firstbase/fb.h>
.sp 1
fb_bulkrec_begin(db, fwait)
.br
fb_database *db;
.br
int fwait;
.sp 1
fb_bulkrec_end(db)
.br
fb_database *db;
.br
.SH DESCRIPTION
As explained in the \fIdatabase\fP(4)
manual page, a database structure contains
all of the elements needed for an entire database, including an array of
field structures, all file names and descriptors, and index information.
Altogether, the array of field structures can hold one complete record.
.PP
The \fIFirstBase\fP routines \fIfb_addrec\fP and \fIfb_delrec\fP
are used to add and delete records one at a time.
However, sometimes it is desirable to add and delete
many records as fast as possible.
The bulk record processing mechanism can be used in these cases.
.PP
The bulk processing of records consists of three steps. First,
\fIfb_bulkrec_begin\fP is called before any processing. This preliminary
call will lock and synchronize the database \fIdb\fP,
and read the database header.
.PP
Then, \fIfb_b_addrec\fP and/or \fIfb_b_delrec\fP are called repeatedly,
as needed.
Finally, \fIfb_bulkrec_end\fP is used to write the header,
synchronize the data files, and unlock the database.
.PP
The flag \fIfwait\fP used in the \fIfb_b_bulkrec_begin\fP mechanism
controls the behavior of the internal lock request. When \fIfwait\fP is set
to \fIFB_WAIT\fP, \fIfb_b_bulkrec_begin\fP will block until the lock is
granted, or until the timeout is reached. However, when \fIfwait\fP is
\fIFB_NOWAIT\fP, then \fIfb_b_bulkrec_begin\fP will return immediately if
the database is locked. In this case, the return status will be FB_ERROR, but
the external short int variable \fIcdb_error\fP will be set to FB_LOCKED_ERROR.
.SH CLIENT/SERVER
The \fIfb_bulkrec_begin\fP and \fIfb_bulkrec_end\fP routines will transparently
make calls to \fIfb_bulkrec_begin_clnt\fP and \fIfb_bulkrec_end_clnt\fP
when the global \fIFirstBase\fP variable \fIcdb_use_rpc\fP is set to one.
See \fIfb_clnt_create\fP(3) for more details.
.SH SEE ALSO
opendb(3), addrec(3), delrec(3), database(4).
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/librec/bulkrec.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: bulkrec.c,v 9.1 2001/02/16 19:34:18 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Bulkrec_sid[] = "@(#) $Id: bulkrec.c,v 9.1 2001/02/16 19:34:18 john Exp $";
#endif

/*
 * these routines enable the use of batch record processing
 */

#include <fb.h>
#include <fb_ext.h>

static int save_locklevel = -1;
extern short int cdb_use_rpc;
extern short int cdb_returnerror;
extern short int cdb_error;
extern short int cdb_locklevel;
extern short int cdb_secure;

/*
 * fb_bulkrec_begin - begin part for any bulk record processes
 */

   fb_bulkrec_begin(dp, fwait)
      fb_database *dp;
      int fwait;

      {
         int st = FB_AOK;

#if RPC
         if (cdb_use_rpc)
            return(fb_bulkrec_begin_clnt(dp, fwait));
#endif /* RPC */
         cdb_error = 0;
         st = fb_lock(0L, dp, fwait);
         if (st == FB_ERROR){
            if (fwait == FB_NOWAIT)
               cdb_error = FB_LOCKED_ERROR;
            else
               cdb_error = FB_ERROR;
            return(FB_ERROR);
            }
	 fb_setdirty(dp, 1);
	 fb_allsync(dp);
         st = fb_gethead(dp);
	 if (st == FB_ERROR){
            cdb_error = FB_READ_ERROR;
	    fb_lerror(FB_READ_ERROR, SYSMSG[S_BAD_HEADER], dp->dbase);
            return(FB_ERROR);
            }
         /*
          * this noise is due to fb_getrec() doing a lock on 0
          *    if the locklevel is 2 ... would cause deadlock here.
          */
         if (save_locklevel == -1)
            save_locklevel = cdb_locklevel;
         cdb_locklevel = 1;
         dp->inside_bulkrec = 1;
         return(st);
      }

/*
 * fb_bulkrec_end - end for any bulkrec processes
 */

   fb_bulkrec_end(dp)
      fb_database *dp;

      {
#if RPC
         if (cdb_use_rpc)
            return(fb_bulkrec_end_clnt(dp));
#endif /* RPC */
	 if (fb_puthead(dp->fd, dp->reccnt, dp->delcnt) == FB_ERROR){
            cdb_error = FB_WRITE_ERROR;
	    fb_lerror(FB_WRITE_ERROR, SYSMSG[S_BAD_HEADER], dp->dbase);
            return(FB_ERROR);
            }
	 fb_setdirty(dp, 0);
	 fb_allsync(dp);
	 fb_unlock_head(dp);
         cdb_locklevel = save_locklevel;
         dp->inside_bulkrec = 0;
         return(FB_AOK);
      }

/* 
 * b_addrec - bulk addrec - no locking or syncs done here.
 *	this is the same as fb_addrec() but no locking.
 */

   fb_b_addrec(dp)
      fb_database *dp;
   
      {
#if RPC
         if (cdb_use_rpc)
            return(fb_b_addrec_clnt(dp));
#endif /* RPC */
         if (dp->inside_bulkrec == 0)
            return(FB_ERROR);
         /* store a blank so that delrec works with no length thrashing */
         if (cdb_secure)
	    fb_recmode(dp, FB_BLANK, fb_getuid(), fb_getgid(), "666");
         else
            fb_store(dp->kp[dp->nfields], SYSMSG[S_STRING_BLANK], dp);
         /*
          * this is a new record; it cannot have previous auto values. clear.
          */
         fb_clear_autoindex(dp);
         fb_checkauto(dp);
         dp->reccnt++;
	 if (fb_putrec(dp->reccnt, dp) == FB_ERROR){
            cdb_error = FB_WRITE_ERROR;
	    fb_lerror(FB_WRITE_ERROR, SYSMSG[S_BAD_DATA], dp->dbase);
            return(FB_ERROR);
            }
         dp->rec = dp->reccnt;
	 if (fb_put_autoindex(dp) == FB_ERROR){
            cdb_error = FB_WRITE_ERROR;
	    fb_lerror(FB_WRITE_ERROR, SYSMSG[S_BAD_INDEX], dp->dbase);
            return(FB_ERROR);
            }
	 return(FB_AOK);
      }

/* 
 * b_delrec - bulk delrec - no locking or syncs done here.
 *	this is the same as fb_delrec() but no locking.
 */

   fb_b_delrec(dp)
      fb_database *dp;
      
      {
         int st = FB_AOK;

#if RPC
         if (cdb_use_rpc)
            return(fb_b_delrec_clnt(dp));
#endif /* RPC */
         if (dp->inside_bulkrec == 0)
            return(FB_ERROR);
         if (cdb_secure)
            fb_recmode(dp, CHAR_STAR, -1, -1, NIL);	/* update del only */
         else{
            /* make sure there is enogh room to store the '*' */
            fb_store(dp->kp[dp->nfields], SYSMSG[S_STRING_BLANK], dp);
            dp->kp[dp->nfields]->fld[0] = CHAR_STAR;
            }
	 if (fb_putrec(dp->rec, dp) == FB_ERROR){
            cdb_error = FB_WRITE_ERROR;
	    fb_lerror(FB_WRITE_ERROR, SYSMSG[S_BAD_DATA], dp->dbase);
            return(FB_ERROR);
            }
         dp->delcnt++;
         if (dp->b_tree == 1 || dp->b_autoindex != NULL)
            st = fb_delidx(dp, dp->rec);
	 return(st);
      }
